### PR TITLE
added p2 poms and repo

### PR DESCRIPTION
**Pull request by kaikreuzer**
This creates a p2 repo that can be included in the target platform of the existing openHAB Add-ons IDE.
I have also included the parent poms required to build addons using Tycho, just to not break anything.

Once all addon repos have been migrated to pure Maven, this whole `features/p2` folder can be removed.

Signed-off-by: Kai Kreuzer <kai@openhab.org>

### CODE REVIEW

**Comment by kaikreuzer** [comment]
@maggu2810 Once you've merged this, I'd think we could go and merge your PR - and fix and enhance everything further from there.

File: features/p2/poms/pom.xml [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+  <licenses>
+    <license>
+      <name>Eclipse Public License</name>
+      <url>http://www.eclipse.org/legal/epl-v20.html</url>
+    </license>
+  </licenses>
+
+  <prerequisites>
+    <maven>3.1</maven>
```
Comment (maggu2810) [inline features/p2/poms/pom.xml]: That prerequisites is for maven plugins. Drop it.

File: features/p2/poms/tycho/pom.xml [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+    <dependency>
+      <groupId>org.openhab.core.bundles</groupId>
+      <artifactId>org.openhab.core.ui.icon</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.openhab.core.bundles</groupId>
+      <artifactId>org.openhab.core.test</artifactId>
+      <version>${project.version}</version>
+    </dependency>
```
Comment (maggu2810) [inline features/p2/poms/tycho/pom.xml]: Have you seen my openhab-core BOM? If there are some openHAB Core bundles missing, you should add them there and this one could depend on that BOM.

File: features/p2/poms/pom.xml (hunk 1 above)
Comment (maggu2810) [reply features/p2/poms/pom.xml]: https://issues.apache.org/jira/browse/MNG-6092

Comment (kaikreuzer) [reply features/p2/poms/pom.xml]: removed.

File: features/p2/poms/tycho/pom.xml (hunk 2 above)
Comment (kaikreuzer) [reply features/p2/poms/tycho/pom.xml]: makes sense, done.

Comment (maggu2810) [reply features/p2/poms/tycho/pom.xml]: Have you checked if all artifacts has been part of the openhab-core BOM?
I just checked some of them. Missing:

* org.openhab.core.ui.dashboard
* org.openhab.core.test
* org.openhab.core.boot
* org.openhab.core.compat1x

It seems I missed the artifacts that has been part of that repo (and not ESH) before.
Can you check them?

**Comment by maggu2810** [comment]
It does not work for me:

```text
$ mvn clean install
[INFO] Scanning for projects...
[ERROR] [ERROR] Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM for org.openhab.core.features.p2.repo:org.openhab.core.features.p2.repo:[unknown-version]: Could not find artifact org.openhab:pom:pom:2.5.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 5, column 11
 @ 
[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.openhab.core.features.p2.repo:org.openhab.core.features.p2.repo:[unknown-version] (/.../openhab-core/features/p2/repo/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for org.openhab.core.features.p2.repo:org.openhab.core.features.p2.repo:[unknown-version]: Could not find artifact org.openhab:pom:pom:2.5.0-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 5, column 11 -> [Help 2]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/ProjectBuildingException
[ERROR] [Help 2] http://cwiki.apache.org/confluence/display/MAVEN/UnresolvableModelException
```

**Comment by maggu2810** [comment]
Used the a small shell command line to identify the missing artifacts in the BOM:

* org.openhab.core.automation.sample.extension.java
* org.openhab.core.automation.sample.extension.json
* org.openhab.core.automation.sample.java.demo
* org.openhab.core.automation.sample.json.demo
* org.openhab.core.automation.sample.moduletype.demo
* org.openhab.core.automation.sample.rest.api
* org.openhab.core.boot
* org.openhab.core.compat1x
* org.openhab.core.io.jetty.certificate
* org.openhab.core.karaf
* org.openhab.core.magic
* org.openhab.core.model.codegen
* org.openhab.core.model.lazygen
* org.openhab.core.test
* org.openhab.core.ui.dashboard

I assume the "sample" and the "...gen" don't need to be present, but the remaining ones should be added.

**Comment by maggu2810** [comment]
Thanks for your work, but it seems there is still something wrong...


To ensure that all dependencies are available in the local Maven repository I built this branch locally once.
After that I removed all openHAB bundles from the repository

```sh
rm -rf ~/.m2/repository/org/openhab/
```

and started a new "offline" build.

All except the last to projects have been able to build.

```text
[INFO] openHAB Core :: Features :: Karaf :: Core .......... SUCCESS [ 45.586 s]
[INFO] openHAB Addons Parent Pom .......................... SUCCESS [  0.059 s]
[INFO] openHAB - Tycho settings ........................... SUCCESS [  0.303 s]
[INFO] openHAB Core P2 Repository ......................... FAILURE [  1.324 s]
[INFO] openHAB Core :: Features :: p2 ..................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  04:33 min
[INFO] Finished at: 2019-01-28T09:58:05+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.reficio:p2-maven-plugin:1.2.0-SNAPSHOT:site (default-cli) on project org.openhab.core.features.p2.repo: Execution default-cli of goal org.reficio:p2-maven-plugin:1.2.0-SNAPSHOT:site failed: org.eclipse.aether.resolution.ArtifactResolutionException: Cannot access openhab-artifactory-snapshot (https://openhab.jfrog.io/openhab/libs-snapshot) in offline mode and the artifact org.openhab.core.bundles:org.openhab.core.automation.sample.extension.java:jar:2.5.0-SNAPSHOT has not been downloaded from it before. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :org.openhab.core.features.p2.repo
```

**Comment by maggu2810** [comment]
The module `org.openhab.core.automation.sample.extension.java` is missing in the `bundles/pom.xml` file

**Comment by kaikreuzer** [comment]
> The module org.openhab.core.automation.sample.extension.java is missing in the bundles/pom.xml file

The samples are meant as documentation and not as build artifacts - I thus don't think they should be included in the build (I'd rather vote for removing them completely...)

**Comment by maggu2810** [comment]
It seems multiple existing modules have been removed from the reactor by `a6072945ea01ad51c035ff9ad9bcd35ba2faace1`

```sh
(cd bundles; ls -d *automation*) | while read MODULE; do grep "<module>${MODULE}</module>" bundles/pom.xml 1>/dev/null || echo "missing: ${MODULE}"; done
```

```
missing: org.openhab.core.automation.sample.extension.java
missing: org.openhab.core.automation.sample.extension.json
missing: org.openhab.core.automation.sample.java.demo
missing: org.openhab.core.automation.sample.json.demo
missing: org.openhab.core.automation.sample.moduletype.demo
missing: org.openhab.core.automation.sample.rest.api
```

**Comment by kaikreuzer** [comment]
Regarding the build error: Are you sure you had pulled https://github.com/maggu2810/openhab-core/pull/5/commits/d513c6d78c93b905a6e4864fcea09aa69b506996 as well?

**Comment by maggu2810** [comment]
The samples should be part of the build (not necessary of the BOM or the p2 feature) similar to the tests (sure, currently a lot of tests are missing) as long as they are part of the repository.
Otherwise we don't detect if the samples are broken and as you stated on the documentation topic yourself: documentation should be up-to-date and not outdated

**Comment by maggu2810** [comment]
> Regarding the build error: Are you sure you had pulled [d513c6d](https://github.com/maggu2810/openhab-core/commit/d513c6d78c93b905a6e4864fcea09aa69b506996) as well?

No, this one is missing.
My working copy has been based on your commit before "fixed magic bundle".

**Comment by maggu2810** [comment]
Okay to remove it. The samples are still available in the ESH repository and we already agreed that the documentation of the framework needs to be improved / re-added.

**Comment by kaikreuzer** [comment]
Ok, perfect - they are removed now :-)

**Pull request merged**